source/val/Function.cpp

```diff
@@ -41,6 +44,53 @@ using std::pair;
 using std::tie;
 using std::vector;
 
+namespace {
+
+using libspirv::BasicBlock;
+
+// Computes a minimal set of root nodes required to traverse, in the forward
+// direction, the CFG represented by the given vector of blocks, and successor
+// and predecessor functions.
+std::vector<BasicBlock*> TraversalRoots(const std::vector<BasicBlock*>& blocks,
+                                        libspirv::get_blocks_func succ_func,
+                                        libspirv::get_blocks_func pred_func) {
+  // The set of nodes which have been visited from any of the roots so far.
+  std::unordered_set<const BasicBlock*> visited;
+
+  auto mark_visited = [&visited](const BasicBlock* b) { visited.insert(b); };
+  auto ignore_block = [](const BasicBlock*) {};
+  auto ignore_blocks = [](const BasicBlock*, const BasicBlock*) {};
+
+  auto traverse_from_root = [&mark_visited, &succ_func, &ignore_block,
+                             &ignore_blocks](const BasicBlock* entry) {
+    DepthFirstTraversal(entry, succ_func, mark_visited, ignore_block,
+                        ignore_blocks);
+  };
+
+  std::vector<BasicBlock*> result;
+
+  // First collect nodes without predecessors.
+  for (auto block : blocks) {
```

Comment:
It would probably be a good idea to perform the reachability propagation using this approach. I don't like how I am performing that now. I can do that in a future PR.